# 📌 Grocery Customer Churn (Abandono de Clientes de Supermercado)

## Acerca del conjunto de datos

Este conjunto de datos contiene información detallada sobre las transacciones, el comportamiento y los datos demográficos de los clientes de una tienda de comestibles. Incluye información de los clientes, datos transaccionales y métricas de comportamiento, lo que lo hace ideal para crear modelos predictivos para la pérdida de clientes y el análisis del valor de vida del cliente (CLV).

La columna objetivo para la predicción de abandono es la tasa de abandono, que indica si un cliente se ha dado de baja (1) o ha permanecido activo (0). El conjunto de datos está diseñado para crear modelos y predecir la retención de clientes, analizar su comportamiento y pronosticar su valor de vida.

Características principales:

* Información del cliente: incluye detalles demográficos como edad, sexo, nivel de ingresos, estado civil, nivel educativo y ocupación.

* Datos transaccionales: contiene información sobre cada transacción, como la fecha de la transacción, la cantidad, el precio, la categoría del producto y el método de pago.

* Métricas de comportamiento del cliente: incluye características como el valor de compra promedio, la frecuencia de compra y la fecha de la última compra.

* Datos promocionales: incluye detalles sobre las promociones, incluido el tipo, la eficacia y el público objetivo.

* Churn: La columna de destino indica si el cliente se ha dado de baja (1) o ha permanecido activo (0).

* Tamaño del conjunto de datos: este conjunto de datos contiene 35.843 filas y 26 atributos (incluida la columna de abandono).

Este estudio incluye las tareas del Data Engineer:

* Obtención y tratamiento de datos a través de una API externa.

* Implementación del proceso ETL (Extracción, Transformación y Carga) para la depuración y organización de la información.

* Desarrollo de visualizaciones clave para detectar patrones y comportamientos relevantes.

* Exploración de los datos (EDA técnico) y elaboración de un reporte con hallazgos significativos, para posterior EDA profundo.

##  Extracción

In [138]:
import pandas as pd
import numpy as np
import os
import warnings
warnings.filterwarnings('ignore')
from google.colab import files

# Abrir un cuadro para subir archivos
uploaded = files.upload()

for nombre_archivo in uploaded.keys():
    df = pd.read_csv(nombre_archivo)


Saving Grocery_Customer_Churn_Data.csv to Grocery_Customer_Churn_Data.csv


## Inspección general del dataset

In [139]:
# Configuraciones de pandas

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:.2f}".format)


In [140]:
# Mostrar las filas del DataFrame

df.head()
df.tail()



customer_id  age  gender income_bracket loyalty_program  \
35838       C1338   48  Female         Medium             Yes   
35839       C1339   26  Female            Low             Yes   
35840       C1340   57    Male           High              No   
35841       C1341   43   Other            Low              No   
35842       C1342   53   Other            Low              No   

       membership_years marital_status  number_of_children education_level  \
35838                 1        Married                   0             PhD   
35839                 5       Divorced                   1      Bachelor's   
35840                 2        Married                   1        Master's   
35841                12       Divorced                   4        Master's   
35842                 8       Divorced                   3        Master's   

       occupation transaction_id transaction_date product_category  quantity  \
35838    Employed          T1338       2023-12-05         Clothing         4   
35839     Student          T1339       2023-12-06        Groceries         4   
35840  Unemployed          T1340       2023-12-07       Home Goods         3   
35841     Student          T1341       2023-12-08         Clothing         3   
35842    Employed          T1342       2023-12-09             Toys         3   

       unit_price  avg_purchase_value purchase_frequency last_purchase_date  \
35838       51.07              100.74              Daily         2023-12-05   
35839      148.23               46.29             Yearly         2023-12-06   
35840      107.40               79.63                NaN         2023-12-07   
35841       61.33                6.36             Weekly         2023-12-08   
35842       17.98              134.11             Yearly         2023-12-09   

       avg_discount_used  online_purchases  in_store_purchases  total_sales  \
35838               5.68                10                   5       247.51   
35839               8.76                 5                   0      1119.07   
35840               6.11                 6                   9       724.01   
35841               7.41                 0                   4      5141.65   
35842              13.52                10                   3        47.96   

       total_transactions  total_items_purchased        promotion_type  \
35838                   8                     63     Seasonal Discount   
35839                  16                     32     Seasonal Discount   
35840                  21                     28               20% Off   
35841                  17                     15               20% Off   
35842                  48                     56  Buy One Get One Free   

      promotion_effectiveness  days_since_last_purchase  churn  
35838                     Low                      -157      0  
35839                     Low                      -158      0  
35840                  Medium                      -159      1  
35841                    High                      -160      0  
35842                     Low                      -161      1

In [141]:
# Atributo del DataFrame
df.shape

(35843, 28)

In [142]:
# Obtener información general del DataFrame

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35843 entries, 0 to 35842
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customer_id               35843 non-null  object 
 1   age                       35843 non-null  int64  
 2   gender                    35843 non-null  object 
 3   income_bracket            35843 non-null  object 
 4   loyalty_program           35843 non-null  object 
 5   membership_years          35843 non-null  int64  
 6   marital_status            35843 non-null  object 
 7   number_of_children        35843 non-null  int64  
 8   education_level           35843 non-null  object 
 9   occupation                35843 non-null  object 
 10  transaction_id            35843 non-null  object 
 11  transaction_date          35843 non-null  object 
 12  product_category          35843 non-null  object 
 13  quantity                  35843 non-null  int64  
 14  unit_p

**Nota:**

El dataset contiene 35.843 registros y 28 columnas, combinando información de clientes y transacciones.

Se observa una mezcla de variables:

* Identificadores (customer_id, transaction_id)

* Variables demográficas del cliente

* Variables transaccionales

* Variables agregadas históricas

Esto indica que el dataset no está normalizado y presenta múltiples niveles de granularidad.

* transaction_date, last_purchase_date → object ❌

* purchase_frequency → debería ser categórica (object está bien, pero validar valores)

* promotion_effectiveness → categórica, no numérica

## Clasificar variables

In [143]:
num_cols = [
    "age", "membership_years", "number_of_children",
    "quantity", "unit_price", "avg_purchase_value",
    "purchase_frequency", "avg_discount_used",
    "online_purchases", "in_store_purchases",
    "total_sales", "total_transactions",
    "total_items_purchased", "promotion_effectiveness",
    "days_since_last_purchase"
]

categorical_cols = [
    "gender", "income_bracket", "loyalty_program",
    "marital_status", "education_level", "occupation",
    "product_category", "promotion_type", "churn"
]

date_cols = [
    "transaction_date", "last_purchase_date"
]

id_cols = [
    "customer_id", "transaction_id"
]


Se realizó una clasificación explícita de columnas por tipo lógico (numéricas, categóricas, fechas e identificadores) con el objetivo de:

* Facilitar validaciones de calidad de datos

* Aplicar transformaciones específicas por tipo

* Mejorar la mantenibilidad del pipeline ETL

## Detección de valores nulos

In [144]:
nulls = df.isna().sum().sort_values(ascending=False)
nulls[nulls > 0]


promotion_type        5213
avg_purchase_value    5206
purchase_frequency    5179
total_sales           3584
dtype: int64

**Nota:**

Se identifican valores faltantes principalmente en variables relacionadas con promociones, frecuencia de compra y métricas agregadas.

Columnas con mayor cantidad de nulos:

* promotion_type: 5213
* avg_purchase_value: 5206
* purchase_frequency: 5179
* total_sales: 3584

Esto sugiere que:

* No todas las transacciones estuvieron asociadas a promociones

* Algunas métricas agregadas pueden no haber sido calculadas para todos los registros

Estos nulos no se consideran errores automáticamente, sino que requieren una decisión de negocio (imputación, categorización explícita o exclusión según el caso).

* promotion_type nulo → no todas las transacciones tuvieron promoción

* purchase_frequency y avg_purchase_value → no estén calculadas para todos los clientes

* total_sales nulo → inconsistencia, posiblemente falte para clientes sin historial consolidado


## Duplicados y unicidad

In [145]:
df["customer_id"].duplicated().sum()


np.int64(35343)

In [146]:
df["transaction_id"].duplicated().sum()


np.int64(35343)

In [147]:
df.groupby("customer_id")["transaction_id"].nunique().describe()


count   500.00
mean      1.00
std       0.00
min       1.00
25%       1.00
50%       1.00
75%       1.00
max       1.00
Name: transaction_id, dtype: float64

In [148]:
df["transaction_id"].nunique()

500

In [149]:
df["customer_id"].value_counts().head()

customer_id
C1342    72
C1341    72
C1340    72
C1339    72
C1338    72
Name: count, dtype: int64

**Nota:**

customer_id duplicados: 35.343

transaction_id:

* count = 500

* mean = 1

* std = 0

Conclusión estructural:

* Cada transaction_id aparece ~72 veces

* Pero cada customer_id aparece muchas veces

* transaction_id NO es una transacción. Además  transaction_id identifica un CUSTOMER, no un evento




## EDA técnico

### Variables Númericas y Fechas

In [150]:
df[num_cols ].describe()

age  membership_years  number_of_children  quantity  unit_price  \
count 35843.00          35843.00            35843.00  35843.00    35843.00   
mean     44.22              7.68                2.07      2.99      103.78   
std      14.94              4.40                1.42      1.40       56.24   
min      18.00              1.00                0.00      1.00        5.16   
25%      32.00              4.00                1.00      2.00       54.21   
50%      45.00              8.00                2.00      3.00      107.23   
75%      57.00             11.00                3.00      4.00      151.90   
max      70.00             15.00                4.00      5.00      199.75   

       avg_purchase_value  avg_discount_used  online_purchases  \
count            30637.00           35843.00          35843.00   
mean                81.95              10.03              5.03   
std                 42.54               2.87              3.16   
min                 -7.21               5.01              0.00   
25%                 43.36               7.52              2.00   
50%                 84.71              10.35              5.00   
75%                118.64              12.47              8.00   
max                164.19              14.96             10.00   

       in_store_purchases  total_sales  total_transactions  \
count            35843.00     32259.00            35843.00   
mean                 4.74      2634.98               28.62   
std                  3.15      1507.68               13.42   
min                  0.00     -1514.41                5.00   
25%                  2.00      1401.95               18.00   
50%                  4.00      2700.31               29.00   
75%                  8.00      3869.30               41.00   
max                 10.00      6600.61               50.00   

       total_items_purchased  days_since_last_purchase  
count               35843.00                  35843.00  
mean                   55.26                    -67.75  
std                    26.36                    144.18  
min                    10.00                   -318.00  
25%                    31.00                   -192.00  
50%                    57.00                    -67.00  
75%                    77.00                     57.00  
max                   100.00                    181.00

In [151]:
# Lista de columnas numéricas a evaluar

# Función para detectar outliers por regla de negocio
def detect_anomalies(df, cols):
    anomalies = {}

    for col in cols:
        if col not in df.columns:
            continue

        # Forzar conversión numérica
        series_raw = df[col]
        series = pd.to_numeric(series_raw, errors="coerce")

        n_missing = series.isnull().sum()
        n_negative = (series < 0).sum()
        n_zero = (series == 0).sum()

        min_val = series.min()
        max_val = series.max()
        mean_val = series.mean()
        std_val = series.std()

        if pd.notna(std_val) and std_val > 0:
            lower_limit = mean_val - 3 * std_val
            upper_limit = mean_val + 3 * std_val
            n_outliers = ((series < lower_limit) | (series > upper_limit)).sum()
        else:
            n_outliers = 0

        anomalies[col] = {
            "missing": int(n_missing),
            "negative": int(n_negative),
            "zero": int(n_zero),
            "min": min_val,
            "max": max_val,
            "mean": mean_val,
            "std": std_val,
            "outliers_3sigma": int(n_outliers)
        }

    return pd.DataFrame(anomalies).T


In [152]:
# Evaluación numéricas
num_anomalies = detect_anomalies(df, num_cols)
display(num_anomalies.sort_values(by="outliers_3sigma", ascending=False))

# Fechas: detectar fechas futuras o inconsistentes
today = pd.Timestamp.today()

date_cols = ["transaction_date", "last_purchase_date"]

for col in date_cols:
    if col in df.columns:
        df[col] = pd.to_datetime(df[col], errors="coerce")


for col in date_cols:
    if col in df.columns:
        n_missing = df[col].isnull().sum()
        n_future = (df[col] > today).sum()
        min_date = df[col].min()
        max_date = df[col].max()

        print(
            f"{col}: "
            f"missing={n_missing}, "
            f"future_dates={n_future}, "
            f"min={min_date}, "
            f"max={max_date}"
        )


missing  negative    zero      min     max    mean  \
age                          0.00      0.00    0.00    18.00   70.00   44.22   
membership_years             0.00      0.00    0.00     1.00   15.00    7.68   
number_of_children           0.00      0.00 6387.00     0.00    4.00    2.07   
quantity                     0.00      0.00    0.00     1.00    5.00    2.99   
unit_price                   0.00      0.00    0.00     5.16  199.75  103.78   
avg_purchase_value        5206.00     11.00    0.00    -7.21  164.19   81.95   
purchase_frequency       35843.00      0.00    0.00      NaN     NaN     NaN   
avg_discount_used            0.00      0.00    0.00     5.01   14.96   10.03   
online_purchases             0.00      0.00 3295.00     0.00   10.00    5.03   
in_store_purchases           0.00      0.00 3440.00     0.00   10.00    4.74   
total_sales               3584.00    973.00    0.00 -1514.41 6600.61 2634.98   
total_transactions           0.00      0.00    0.00     5.00   50.00   28.62   
total_items_purchased        0.00      0.00    0.00    10.00  100.00   55.26   
promotion_effectiveness  35843.00      0.00    0.00      NaN     NaN     NaN   
days_since_last_purchase     0.00  22739.00   72.00  -318.00  181.00  -67.75   

                             std  outliers_3sigma  
age                        14.94             0.00  
membership_years            4.40             0.00  
number_of_children          1.42             0.00  
quantity                    1.40             0.00  
unit_price                 56.24             0.00  
avg_purchase_value         42.54             0.00  
purchase_frequency           NaN             0.00  
avg_discount_used           2.87             0.00  
online_purchases            3.16             0.00  
in_store_purchases          3.15             0.00  
total_sales              1507.68             0.00  
total_transactions         13.42             0.00  
total_items_purchased      26.36             0.00  
promotion_effectiveness      NaN             0.00  
days_since_last_purchase  144.18             0.00

transaction_date: missing=0, future_dates=0, min=2023-01-01 00:00:00, max=2024-05-14 00:00:00
last_purchase_date: missing=0, future_dates=0, min=2023-01-01 00:00:00, max=2024-05-14 00:00:00


**Nota:**

1️⃣ Columnas numéricas

* age, membership_years, number_of_children, quantity, unit_price

Sin negativos, sin outliers según 3σ → ✅ saneadas

* avg_purchase_value

5.206 nulos → ya identificado

Min = -7,21 → valores negativos pequeños, probablemente errores de transacción

Outliers 3σ = 0 → estadísticamente dentro de rango

* purchase_frequency

Todos nulos → ❌ requiere recalcular

* total_sales

3.584 nulos → ya identificado

973 negativos → ❌ valores problemáticos

Min = -1.514 → claramente errores o cálculos incorrectos

* online_purchases, in_store_purchases

Muchos ceros (≈ 3.300-3.400) → puede ser normal según reglas de negocio

* promotion_effectiveness

Todos nulos → ❌ descartar o recalcular

* days_since_last_purchase

22.739 negativos → ❌ problema sistémico confirmado

2️⃣ Columnas de fecha

* transaction_date y last_purchase_date

No hay nulos, no hay fechas futuras → ✅ consistentes

Observación: el gran problema de fechas negativas ya no es por fechas inválidas, sino por cálculo incorrecto de days_since_last_purchase → debe recalcularse como (hoy - last_purchase_date)

3️⃣ Conclusiones parciales

Las fechas son confiables, no hay que imputar ni corregir transaction_date ni last_purchase_date.

days_since_last_purchase debe recalcularse para eliminar los negativos.

purchase_frequency y promotion_effectiveness no se pueden usar tal cual, requieren recalculo o imputación.

avg_purchase_value y total_sales tienen algunos valores negativos. Revisar cómo se calcularon (probablemente quantity * unit_price para total_sales).

El resto de variables numéricas están limpias y listas para análisis o derivación de features.

In [153]:
(df["age"] < 0).sum(), (df["age"] > 100).sum()


(np.int64(0), np.int64(0))

In [154]:
(df["unit_price"] < 0).sum()


np.int64(0)

In [155]:
(df["quantity"] <= 0).sum()


np.int64(0)

In [156]:
(df["transaction_date"] > pd.Timestamp.today()).sum()


np.int64(0)

In [157]:
# Detección de anomalías lógicas

anomalies = {}

# Ventas negativas
anomalies["negative_total_sales"] = df[df["total_sales"] < 0].shape[0]

# Cantidades inválidas
anomalies["invalid_quantity"] = df[df["quantity"] <= 0].shape[0]

# Precios inválidos
anomalies["invalid_unit_price"] = df[df["unit_price"] <= 0].shape[0]

# Edades fuera de rango lógico
anomalies["invalid_age"] = df[(df["age"] < 18) | (df["age"] > 100)].shape[0]

# Métricas temporales inconsistentes
if "days_since_last_purchase" in df.columns:
    anomalies["negative_days_since_last_purchase"] = (
        df[df["days_since_last_purchase"] < 0].shape[0]
    )

# Resumen de anomalías
anomaly_df = pd.DataFrame.from_dict(
    anomalies, orient="index", columns=["count"]
)

print("\nResumen de anomalías detectadas:")
display(anomaly_df)


Resumen de anomalías detectadas:


count
negative_total_sales                 973
invalid_quantity                       0
invalid_unit_price                     0
invalid_age                            0
negative_days_since_last_purchase  22739

**Notas:**

El análisis de calidad de datos identificó dos tipos principales de anomalías:

* Ventas totales negativas (negative_total_sales):
Se detectaron 973 registros con valores negativos, lo que puede estar asociado a devoluciones, cancelaciones o ajustes contables. Estos casos requieren validación de negocio para decidir si deben mantenerse, corregirse o excluirse.

* Días negativos desde la última compra (negative_days_since_last_purchase):
Se identificaron 22.739 registros con valores negativos, lo que indica inconsistencias temporales (por ejemplo, fechas de referencia mal definidas o eventos posteriores al “snapshot” de análisis). Esta anomalía sugiere un problema de alineación temporal más que de valores numéricos.

* No se detectaron anomalías en:

1.Cantidades (invalid_quantity)

2.Precios unitarios (invalid_unit_price)

3.Edad de clientes (invalid_age)

### Variables Catégoricas

In [158]:
for col in categorical_cols:
    print("\n", col)
    display(df[col].value_counts(dropna=False).head(10))



 gender


gender
Male      12116
Other     12112
Female    11615
Name: count, dtype: int64


 income_bracket


income_bracket
High      13769
Low       11746
Medium    10328
Name: count, dtype: int64


 loyalty_program


loyalty_program
Yes    18844
No     16999
Name: count, dtype: int64


 marital_status


marital_status
Single      13182
Divorced    11333
Married     11328
Name: count, dtype: int64


 education_level


education_level
Master's       9825
PhD            9108
Bachelor's     8810
High School    8100
Name: count, dtype: int64


 occupation


occupation
Employed         7812
Self-Employed    7750
Unemployed       7239
Retired          6879
Student          6163
Name: count, dtype: int64


 product_category


product_category
Home Goods     8022
Clothing       7608
Toys           7308
Electronics    6885
Groceries      6020
Name: count, dtype: int64


 promotion_type


promotion_type
Seasonal Discount       11036
20% Off                  9961
Buy One Get One Free     9633
NaN                      5213
Name: count, dtype: int64


 churn


churn
0    25187
1    10656
Name: count, dtype: int64

**Notas:**

1️⃣ Género

Male: 12.116

Other: 12.112

Female: 11.615

✅ Hay una distribución bastante equilibrada entre géneros, con una ligera ventaja de “Male” y “Other” sobre “Female”. Esto es útil, porque no hay un sesgo fuerte hacia un género en los análisis posteriores.

2️⃣ Income Bracket (Nivel de ingreso)

High: 13.769

Low: 11.746

Medium: 10.328

✅ Predomina el nivel High, seguido de Low y Medium. Esto indica que la base de clientes tiene una proporción relativamente alta de ingresos altos, lo que puede influir en patrones de consumo y promociones.

3️⃣ Loyalty Program

Yes: 18.844

No: 16.999

✅ La mayoría de los clientes participa en un programa de fidelidad, pero hay una proporción significativa que no lo hace. Esto es clave para análisis de retención o churn.

4️⃣ Marital Status

Single: 13.182

Divorced: 11.333

Married: 11.328

✅ Hay más clientes solteros, aunque los divorciados y casados están bastante equilibrados. Esto puede impactar en patrones de compra según la etapa de vida.

5️⃣ Education Level

Master’s: 9.825

PhD: 9.108

Bachelor’s: 8.810

High School: 8.100

✅ La mayoría tiene educación avanzada, con predominio de Master’s y PhD. Esto podría correlacionarse con ingresos altos y ciertos hábitos de consumo.

6️⃣ Occupation

Employed: 7.812

Self-Employed: 7.750

Unemployed: 7.239

Retired: 6.879

Student: 6.163

✅ Hay una diversidad de ocupaciones, siendo la mayoría empleados o autónomos. Esto puede afectar frecuencia de compra y preferencias de productos.

7️⃣ Product Category

Home Goods: 8.022

Clothing: 7.608

Toys: 7.308

Electronics: 6.885

Groceries: 6.020

✅ Los productos más comprados son Home Goods y Clothing, mientras que Groceries es la categoría menos frecuente en este dataset, lo que puede reflejar un enfoque en bienes duraderos o de consumo discrecional.

8️⃣ Promotion Type

Seasonal Discount: 11.036

20% Off: 9.961

Buy One Get One Free: 9.633

NaN: 5.213

✅ La mayoría de las transacciones tienen algún tipo de promoción aplicada, aunque hay un 14–15% de valores faltantes. Esto indica que las promociones son un factor importante para analizar su efecto en ventas y churn.



Conclusión general

* La base de clientes es bastante equilibrada en género y estado civil, aunque ligeramente inclinada hacia solteros y hombres/other.

* Hay predominio de ingresos altos y educación avanzada, lo que podría afectar patrones de gasto.

* La participación en programas de fidelidad es alta, lo que permite analizar retención y efectividad de promociones.

* Los productos más populares son Home Goods y Clothing, y la mayoría de las compras incluyen alguna promoción.

* Los datos son en general balanceados, pero hay algunos valores faltantes en promotion_type, avg_purchase_value y otras variables que deberíamos considerar antes de modelar.

## Limpieza numérica y nulos

In [159]:
df_clean = df.copy()

In [160]:
df_clean = df.copy()

df_clean["gender"] = df_clean["gender"].fillna("unknown")

df_clean["age"] = df_clean["age"].clip(0, 100)
df_clean["age"] = df_clean["age"].fillna(df_clean["age"].median())

df_clean["unit_price"] = df_clean["unit_price"].clip(lower=0)


In [161]:
# Normalizar nombres de columnas
def clean_column_names(df_clean):
    df_clean = df.copy()
    df_clean.columns = (
        df.columns
        .str.strip()                # Quitar espacios al inicio/final
        .str.lower()                # Pasar todo a minúsculas
        .str.replace(" ", "_")      # Reemplazar espacios por _
        .str.replace("-", "_")      # Reemplazar guiones por _
        .str.replace(r"[^\w_]", "", regex=True)  # Quitar caracteres especiales
    )
    return df_clean

## Convertir fechas

In [162]:
# Convertir fechas
date_cols = ["transaction_date", "last_purchase_date"]
for col in date_cols:
    df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")

today = pd.Timestamp.today().normalize()

# Detectar y reportar fechas inconsistentes
for col in date_cols:
    n_missing = df[col].isnull().sum()
    n_future = (df[col] > today).sum()
    print(f"{col}: missing={n_missing}, future_dates={n_future}")

# Recalcular métrica correctamente
df_clean["days_since_last_purchase"] = (
    today - df_clean["last_purchase_date"]
).dt.days

# Seguridad final
df_clean.loc[
    df_clean["days_since_last_purchase"] < 0,
    "days_since_last_purchase"
] = 0

transaction_date: missing=0, future_dates=0
last_purchase_date: missing=0, future_dates=0


## Normalización de categorías y reducción de cardinalidad

In [163]:
for col in categorical_cols:

    # Reducir categorías raras (<1% frecuencia)
    freq = df_clean[col].value_counts(normalize=True)
    rare_labels = freq[freq < 0.01].index
    df_clean[col] = df_clean[col].replace(rare_labels, "other")

# Promotion type: rellenar NaN
df_clean["promotion_type"] = df_clean["promotion_type"].fillna("no_promotion")


In [164]:
df_clean["loyalty_program"] = df_clean["loyalty_program"].astype(str).str.strip().str.lower()

# Mapear Yes/No a 1/0
df_clean["loyalty_program"] = df_clean["loyalty_program"].map({"yes": 1, "no": 0})


## Métricas derivadas

In [165]:
df_clean["total_sales"] = df_clean["quantity"] * df_clean["unit_price"]
df_clean["avg_purchase_value"] = df_clean.groupby("customer_id")["total_sales"].transform("mean")
df_clean["purchase_frequency"] = df_clean.groupby("customer_id")["transaction_id"].transform("count") / df_clean["membership_years"].replace(0,1)

# Promotion effectiveness
df_clean["promo_flag"] = df_clean["promotion_type"].apply(lambda x: 0 if x=="no_promotion" else 1)
df_clean["promotion_effectiveness"] = df_clean.groupby("customer_id")["promo_flag"].transform("sum") / df_clean.groupby("customer_id")["transaction_id"].transform("count")

# Online vs in-store ratio
df_clean["total_purchases"] = df_clean["online_purchases"] + df_clean["in_store_purchases"]
df_clean["online_ratio"] = df_clean["online_purchases"] / df_clean["total_purchases"].replace(0,1)


## Validación post-transformación

In [166]:
for col in num_cols:
    if col in df_clean.columns:
        print(f"{col}: min={df_clean[col].min()}, max={df_clean[col].max()}, mean={df_clean[col].mean():.2f}, nulos={df_clean[col].isnull().sum()}")

age: min=18, max=70, mean=44.22, nulos=0
membership_years: min=1, max=15, mean=7.68, nulos=0
number_of_children: min=0, max=4, mean=2.07, nulos=0
quantity: min=1, max=5, mean=2.99, nulos=0
unit_price: min=5.159564146735754, max=199.7514692043265, mean=103.78, nulos=0
avg_purchase_value: min=6.0056146087811895, max=990.1493097576406, mean=310.11, nulos=0
purchase_frequency: min=4.733333333333333, max=72.0, mean=18.12, nulos=0
avg_discount_used: min=5.012745373771642, max=14.96257824333469, mean=10.03, nulos=0
online_purchases: min=0, max=10, mean=5.03, nulos=0
in_store_purchases: min=0, max=10, mean=4.74, nulos=0
total_sales: min=6.0056146087811895, max=990.1493097576405, mean=310.11, nulos=0
total_transactions: min=5, max=50, mean=28.62, nulos=0
total_items_purchased: min=10, max=100, mean=55.26, nulos=0
promotion_effectiveness: min=0.7183098591549296, max=0.9859154929577465, mean=0.85, nulos=0
days_since_last_purchase: min=594, max=1093, mean=844.25, nulos=0


In [167]:
# Dataset final con todas las columnas procesadas
df_final = df_clean
df_final = df_final.reset_index(drop=True)
df_final.to_csv("dataset_final_all_columns.csv", index=False)

## Dataset Data Analyst

In [168]:
# Dataset resumido por cliente

df_agg = df_final.groupby("customer_id").agg({
    # Numéricas: agregaciones típicas
    "age": "first",
    "membership_years": "first",
    "number_of_children": "first",
    "quantity": "mean",
    "unit_price": "mean",
    "avg_purchase_value": "mean",
    "purchase_frequency": "mean",
    "avg_discount_used": "mean",
    "online_purchases": "sum",
    "in_store_purchases": "sum",
    "total_sales": "sum",
    "total_transactions": "sum",
    "total_items_purchased": "sum",
    "promotion_effectiveness": "mean",
    "days_since_last_purchase": "min",

    # Categóricas: valor más frecuente
    "gender": lambda x: x.mode()[0],
    "income_bracket": lambda x: x.mode()[0],
    "marital_status": lambda x: x.mode()[0],
    "education_level": lambda x: x.mode()[0],
    "occupation": lambda x: x.mode()[0],
    "product_category": lambda x: x.mode()[0],
    "promotion_type": lambda x: x.mode()[0],


    # Fechas
    "transaction_date": "max",
    "last_purchase_date": "max",

    # Flags binarios
    "loyalty_program": "max",
    "churn": "max",
    "promo_flag": "max",

}).reset_index()


# Exportar CSV para Data Analyst

df_agg.to_csv("dataset_analyst_by_customer.csv", index=False)


**Documentación para Data analyst:**


* Diccionario de columnas con descripción y tipo.

* Rango y estadística clave de cada variable.

* Notas sobre imputaciones, clipping y derivadas.

## Dataset Data Scientist

### Codificación de variables categóricas: One-hot encoding

In [187]:
categorical_cols = [
    "gender", "income_bracket", "marital_status",
    "education_level", "occupation",
    "product_category", "promotion_type"
]

df_final_encoded = pd.get_dummies(df_final, columns=categorical_cols, drop_first=False)

In [188]:
# Dataset final para Data Scientist
df_final_encoded.to_csv("customer_dataset_for_ml.csv", index=False)

**Documentación para Data Scientist:**


* Diccionario de columnas codificadas (qué representa cada columna 0/1).

* Notas sobre limpieza y derivadas.

* Estadísticas resumidas de features numéricas.

## Validación final

### 1️⃣ Chequeo rápido de nulos / NaN

In [171]:
def check_nulls(df, name="dataset"):
    print(f"\n📌 Null check → {name}")
    nulls = df.isna().sum()
    nulls = nulls[nulls > 0]

    if nulls.empty:
        print("✅ No hay valores nulos ni NaN")
    else:
        print("❌ Columnas con nulos:")
        print(nulls)


In [172]:
check_nulls(df_final, "df_final")
check_nulls(df_agg, "df_agg")



📌 Null check → df_final
✅ No hay valores nulos ni NaN

📌 Null check → df_agg
✅ No hay valores nulos ni NaN


### 2️⃣Verificar columnas binarias (0/1)

In [173]:
def check_binary_cols(df, binary_cols):
    print("\n📌 Binary columns check")
    for col in binary_cols:
        if col in df.columns:
            invalid = df[~df[col].isin([0,1]) & df[col].notna()]
            if len(invalid) == 0:
                print(f"✅ {col}: OK (solo 0/1)")
            else:
                print(f"❌ {col}: valores inválidos →")
                print(df[col].value_counts())


In [174]:
binary_cols = ["loyalty_program", "churn", "promo_flag" ]
check_binary_cols(df_final, binary_cols)



📌 Binary columns check
✅ loyalty_program: OK (solo 0/1)
✅ churn: OK (solo 0/1)
✅ promo_flag: OK (solo 0/1)


### 3️⃣ Verificación de rangos numéricos

In [175]:
def check_numeric_ranges(df, num_cols):
    print("\n📌 Numeric ranges check")
    for col in num_cols:
        if col in df.columns:
            min_v = df[col].min()
            max_v = df[col].max()
            if min_v < 0:
                print(f"⚠️ {col}: valor negativo detectado (min={min_v})")
            else:
                print(f"✅ {col}: min={min_v}, max={max_v}")


In [176]:
check_numeric_ranges(df_final, num_cols)



📌 Numeric ranges check
✅ age: min=18, max=70
✅ membership_years: min=1, max=15
✅ number_of_children: min=0, max=4
✅ quantity: min=1, max=5
✅ unit_price: min=5.159564146735754, max=199.7514692043265
✅ avg_purchase_value: min=6.0056146087811895, max=990.1493097576406
✅ purchase_frequency: min=4.733333333333333, max=72.0
✅ avg_discount_used: min=5.012745373771642, max=14.96257824333469
✅ online_purchases: min=0, max=10
✅ in_store_purchases: min=0, max=10
✅ total_sales: min=6.0056146087811895, max=990.1493097576405
✅ total_transactions: min=5, max=50
✅ total_items_purchased: min=10, max=100
✅ promotion_effectiveness: min=0.7183098591549296, max=0.9859154929577465
✅ days_since_last_purchase: min=594, max=1093


### 4️⃣ Chequeo de fechas (consistencia temporal)

In [177]:
def check_dates(df, date_cols):
    print("\n📌 Date columns check")
    for col in date_cols:
        if col in df.columns:
            if not pd.api.types.is_datetime64_any_dtype(df[col]):
                print(f"❌ {col}: no es datetime")
            else:
                print(f"✅ {col}: OK ({df[col].min()} → {df[col].max()})")


In [178]:
check_dates(df_final, date_cols)



📌 Date columns check
✅ transaction_date: OK (2023-01-01 00:00:00 → 2024-05-14 00:00:00)
✅ last_purchase_date: OK (2023-01-01 00:00:00 → 2024-05-14 00:00:00)


### 5️⃣ Valores categóricos raros (cardinalidad)

In [179]:
def check_categorical_cols(df, cat_cols, max_unique=20):
    print("\n📌 Categorical consistency check")
    for col in cat_cols:
        if col in df.columns:
            uniq = df[col].nunique(dropna=False)
            print(f"🔹 {col}: {uniq} valores únicos")
            if uniq <= max_unique:
                print(df[col].value_counts())


In [180]:
check_categorical_cols(df_agg, categorical_cols)



📌 Categorical consistency check
🔹 gender: 3 valores únicos
gender
Other     169
Male      169
Female    162
Name: count, dtype: int64
🔹 income_bracket: 3 valores únicos
income_bracket
High      192
Low       164
Medium    144
Name: count, dtype: int64
🔹 marital_status: 3 valores únicos
marital_status
Single      184
Divorced    158
Married     158
Name: count, dtype: int64
🔹 education_level: 4 valores únicos
education_level
Master's       137
PhD            127
Bachelor's     123
High School    113
Name: count, dtype: int64
🔹 occupation: 5 valores únicos
occupation
Employed         109
Self-Employed    108
Unemployed       101
Retired           96
Student           86
Name: count, dtype: int64
🔹 product_category: 5 valores únicos
product_category
Home Goods     112
Clothing       106
Toys           102
Electronics     96
Groceries       84
Name: count, dtype: int64
🔹 promotion_type: 3 valores únicos
promotion_type
Seasonal Discount       180
20% Off                 163
Buy One Get One

### 6️⃣ Chequeo de duplicados

In [181]:
def check_duplicates(df, subset=None, name="dataset"):
    print(f"\n📌 Duplicate check → {name}")
    dups = df.duplicated(subset=subset).sum()
    if dups == 0:
        print("✅ No hay duplicados")
    else:
        print(f"❌ {dups} duplicados encontrados")


In [182]:
check_duplicates(df_agg, subset=["customer_id"], name="df_agg")



📌 Duplicate check → df_agg
✅ No hay duplicados


### 7️⃣ Check FINAL tipo “semáforo”

In [183]:
def final_dataset_check(df, name):
    print(f"\n🚦 FINAL CHECK → {name}")
    print("Filas:", df.shape[0])
    print("Columnas:", df.shape[1])
    print("NaN totales:", df.isna().sum().sum())


In [184]:
final_dataset_check(df_final, "Dataset Final (All Columns)")
final_dataset_check(df_agg, "Dataset Analyst by Customer")



🚦 FINAL CHECK → Dataset Final (All Columns)
Filas: 35843
Columnas: 31
NaN totales: 0

🚦 FINAL CHECK → Dataset Analyst by Customer
Filas: 500
Columnas: 28
NaN totales: 0


##